src/olm/session/mod.rs

```diff
@@ -599,6 +604,53 @@ mod test {
         Ok(())
     }
 
+    #[test]
+    fn max_keys_out_of_order_decryption() -> Result<()> {
+        let (_, _, mut alice_session, bob_session) = sessions()?;
+
+        let mut messages: Vec<messages::OlmMessage> = Vec::new();
+        for i in 0..(MAX_MESSAGE_KEYS + 2) {
+            messages.push(bob_session.encrypt(format!("Message {}", i).as_str()).into());
+        }
+
+        // Decrypt last message
+        assert_eq!(
+            format!("Message {}", MAX_MESSAGE_KEYS + 1).as_bytes(),
+            alice_session.decrypt(&messages[MAX_MESSAGE_KEYS + 1])?
+        );
+
+        // Cannot decrypt first message because it is more than MAX_MESSAGE_KEYS ago
+        assert_matches!(
+            alice_session.decrypt(&messages[0]),
+            Err(DecryptionError::MissingMessageKey(_))
+        );
+
+        // Can decrypt all other messages
+        #[allow(clippy::needless_range_loop)]
```

Comment:
I believe this isn't a needless range loop because I'm iterating over a subset of the vector only

Reply:
What about the suggestion clippy gives us?

```rust
for (i, message) in messages.iter().enumerate().skip(1).take(MAX_MESSAGE_KEYS) {
    assert_eq!(format!("Message {}", i).as_bytes(), alice_session.decrypt(message)?);
}
```

Seems to be a sensible suggestion and you're still iterating over the subset.

Reply:
Sorry, completely missed that. 🤦‍♂️ Now fixed.